# Learning Dataframes

This is me, thinking out loud, after/while watching a few Julia videos.

Julia looks a lot like Python. In fact, there might be a competition somewhere about who can write the longest program executable in both languages. Nah, a few colons and details might give it away though.

Importing packages, we already talked about this, no reason to get all excited about it.

In [2]:
using Plots, DataFrames, CSV, Statistics, Pipe

We have a small file named `Batch_4154922_batch_results.csv` which is the output of an MTurk job, queued up for this. The CSV library looks to be the industry standard for reading CSVs. That kinda makes sense, when you think about it.

Also: Look at the cool pipe. `df` gets its dataframe content by piping (the weird `|>` symbol) the CSV File into the DataFrame function.

In [3]:
df = CSV.File("Batch_4154922_batch_results.csv", silencewarnings=true) |> DataFrame

1760×31 DataFrame. Omitted printing of 29 columns
│ Row  │ HITId                          │ HITTypeId                      │
│      │ String                         │ String                         │
├──────┼────────────────────────────────┼────────────────────────────────┤
│ 1    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 2    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 3    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 4    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 5    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 6    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 7    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 8    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 9    │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 10   │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
⋮
│ 1750 │ 3ZXV7Q5FJANM9OM3ZG2BY0VF5SPCF5 │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1751 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1752 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1753 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1754 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1755 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1756 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1757 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1758 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1759 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 1760 │ 3ZZAYRN1I5QX1I17K5GRA44XLFGOTM │ 36ENCJ709J8WC06EPA4JOHGF537KEH │

Some warnings there so we had to use `silencewarnings=true` to not get 1760 warnings printed. We don't want that. But let's worry about those later. So far, it looks like we have a dataframe. There is a `head` function but it told me it was deprecated and that I should use `first` instead.

In [4]:
first(df, 5)

5×31 DataFrame. Omitted printing of 29 columns
│ Row │ HITId                          │ HITTypeId                      │
│     │ String                         │ String                         │
├─────┼────────────────────────────────┼────────────────────────────────┤
│ 1   │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 2   │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 3   │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 4   │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │
│ 5   │ 301KG0KX9BKTP2L7XD15P8Y69GP2HK │ 36ENCJ709J8WC06EPA4JOHGF537KEH │

Somehow this uses only the first three columns and there doesn't seem to be a way to get more columns.

Also, I learned you could do this:

In [5]:
?first

search: first firstindex popfirst! pushfirst! uppercasefirst lowercasefirst



```
first(coll)
```

Get the first element of an iterable collection. Return the start point of an [`AbstractRange`](@ref) even if it is empty.

# Examples

```jldoctest
julia> first(2:2:10)
2

julia> first([1; 2; 3; 4])
1
```

---

```
first(s::AbstractString, n::Integer)
```

Get a string consisting of the first `n` characters of `s`.

```jldoctest
julia> first("∀ϵ≠0: ϵ²>0", 0)
""

julia> first("∀ϵ≠0: ϵ²>0", 1)
"∀"

julia> first("∀ϵ≠0: ϵ²>0", 3)
"∀ϵ≠"
```

---

```
first(q::Deque)
```

Returns the first element of the deque `q`.

---

```
first(D::CircularDeque)
```

Get the item at the front of the queue.

---

```
first(s::Stack)
```

Get the top item from the stack. Sometimes called peek.

---

```
first(h::BinaryHeap)
```

Returns the element at the top of the heap `h`.

---

```
first(h::BinaryMinMaxHeap)
```

Get the first (minimum) of the heap.

---

```
first(sc)
```

Argument `sc` is a SortedDict, SortedMultiDict or SortedSet. This function returns the first item (a `k=>v` pair for SortedDict and SortedMultiDict or a key for SortedSet) according to the sorted order in the container. Thus, `first(sc)` is equivalent to `deref((sc,startof(sc)))`. It is an error to call this function on an empty container. Time: O(log *n*)

---

```
first(sc)
```

Argument `sc` is a SortedDict, SortedMultiDict or SortedSet. This function returns the first item (a `k=>v` pair for SortedDict and SortedMultiDict or a key for SortedSet) according to the sorted order in the container. Thus, `first(sc)` is equivalent to `deref((sc,startof(sc)))`. It is an error to call this function on an empty container. Time: O(log *n*)

---

```
first(sc)
```

Argument `sc` is a SortedDict, SortedMultiDict or SortedSet. This function returns the first item (a `k=>v` pair for SortedDict and SortedMultiDict or a key for SortedSet) according to the sorted order in the container. Thus, `first(sc)` is equivalent to `deref((sc,startof(sc)))`. It is an error to call this function on an empty container. Time: O(log *n*)

---

```
first(cb::CircularBuffer)
```

Get the first element of CircularBuffer.

---

```
first(df::AbstractDataFrame)
```

Get the first row of `df` as a `DataFrameRow`.

---

```
first(df::AbstractDataFrame, n::Integer)
```

Get a data frame with the `n` first rows of `df`.


Column names are pretty straight forward, but we still don't have a lot of info on what this really is.

In [6]:
names(df)

31-element Array{String,1}:
 "HITId"
 "HITTypeId"
 "Title"
 "Description"
 "Keywords"
 "Reward"
 "CreationTime"
 "MaxAssignments"
 "RequesterAnnotation"
 "AssignmentDurationInSeconds"
 "AutoApprovalDelayInSeconds"
 "Expiration"
 "NumberOfSimilarHITs"
 ⋮
 "AutoApprovalTime"
 "ApprovalTime"
 "RejectionTime"
 "RequesterFeedback"
 "WorkTimeInSeconds"
 "LifetimeApprovalRate"
 "Last30DaysApprovalRate"
 "Last7DaysApprovalRate"
 "Input.document_url"
 "Answer.category.label"
 "Approve"
 "Reject"

In [7]:
summary(df)

"1760×31 DataFrame"

OK that wasn't what I wanted, maybe it was `describe`.

In [8]:
describe(df)

31×8 DataFrame. Omitted printing of 6 columns
│ Row │ variable                    │ mean    │
│     │ Symbol                      │ Union…  │
├─────┼─────────────────────────────┼─────────┤
│ 1   │ HITId                       │         │
│ 2   │ HITTypeId                   │         │
│ 3   │ Title                       │         │
│ 4   │ Description                 │         │
│ 5   │ Keywords                    │         │
│ 6   │ Reward                      │         │
│ 7   │ CreationTime                │         │
│ 8   │ MaxAssignments              │ 10.0    │
│ 9   │ RequesterAnnotation         │         │
│ 10  │ AssignmentDurationInSeconds │ 120.0   │
⋮
│ 21  │ ApprovalTime                │         │
│ 22  │ RejectionTime               │         │
│ 23  │ RequesterFeedback           │         │
│ 24  │ WorkTimeInSeconds           │ 28.6028 │
│ 25  │ LifetimeApprovalRate        │         │
│ 26  │ Last30DaysApprovalRate      │         │
│ 27  │ Last7DaysApprovalRate       │         │
│ 28  │ Input.document_url          │         │
│ 29  │ Answer.category.label       │         │
│ 30  │ Approve                     │         │
│ 31  │ Reject                      │         │

Better. We're still nowhere near where we need to be, but at least we now have a feeling for what is categorical and what is numeric, which is good enough. I want to group and summarize, the de-facto "hello, world" of data. And I heard rumor that there is a pipe operator that is pretty sweet.

In [9]:
@pipe df |> 
    groupby(_, :LifetimeApprovalRate) |> 
    combine(_, :WorkTimeInSeconds => mean, :LifetimeApprovalRate => length)

6×3 DataFrame. Omitted printing of 1 columns
│ Row │ LifetimeApprovalRate │ WorkTimeInSeconds_mean │
│     │ String               │ Float64                │
├─────┼──────────────────────┼────────────────────────┤
│ 1   │ 100% (5/5)           │ 9.232                  │
│ 2   │ 100% (1/1)           │ 28.3122                │
│ 3   │ 100% (2/2)           │ 30.8978                │
│ 4   │ 100% (3/3)           │ 31.6377                │
│ 5   │ 100% (4/4)           │ 23.9587                │
│ 6   │ 0% (0/0)             │ 33.7305                │

Works. Two surprises: 
- the `@pipe` decorator (I'm assuming that's what it is) is less elegant than R
- having to do a count using `length` is... somehow both unnatural and very logical. I hope I'm not doing it right.

That being said, we have now performed the "Hello, World" of data. Let's get down to more business.

Next things do do:
- I want to regex the shit out of some of these variables.
- Not a lot of fan of the column names after aggregation

It seems substring works almost like in Python, except leaving it blank on one side of the colon gives weird results. I guess that's a basic function of how stuff/arrays work in Julia.

In [10]:
"Hello"[:4]

'l': ASCII/Unicode U+006C (category Ll: Letter, lowercase)

In [11]:
"Hello"[1:4]

"Hell"

In [12]:
length("Hello")

5

I have no idea what is going on with negative indexes. It's definetely not like python, and why it doesn't give an error on `"Hello"[2:-2]` is beyond me.

In [13]:
print("Hello"[-2])

BoundsError: BoundsError: attempt to access String
  at index [-2]

In [14]:
print("Hello"[-2:2])

BoundsError: BoundsError: attempt to access String
  at index [-2:2]

In [15]:
print("Hello"[2:-2])

It seems there are some other functions that can help.

In [16]:
print(firstindex("Hello"), ", ", lastindex("Hello"))

1, 5

In [17]:
SubString("Hello", 2, 4)

"ell"

In [18]:
"Hello"[end-1:end]

"lo"

So... this is fairly unusual: There is a special value, `end`, that seems to denote end of string, and that can be used as a helper when subsetting. And there is a `SubString` function that takes takes start-position and end-position (my least favorite arguments). The documentation suggest substring might be a good friend once you venture outside of ASCII territory, because the subsetting is a little too simple-minded. I know it's a detour, but I want to try it:

In [19]:
a = "🐰, how are you?"

println(a[1:5])

🐰,


OK, so the first four characters are a rabbit, which makes sense. Last try:

In [20]:
b = " Hi ÆØÅ"

println(b[1:6])

StringIndexError: StringIndexError(" Hi ÆØÅ", 6)

In [21]:
println(SubString(b, 1, 6))

StringIndexError: StringIndexError(" Hi ÆØÅ", 6)

In [22]:
println(SubString(b, 1, 5))

 Hi Æ


That's the spirit! Everything crashes if you slice an overlong byte. I'm sure there's a way to get around it.

So, to the regex stuff. There doesn't seem to be any separate regex library like python's `re`, and regex-patterns can be defined suspiciously simply:

In [23]:
ptn = r"^[0-9]{1,3}\%"

r"^[0-9]{1,3}\%"

In [24]:
m = match(ptn, "100% (2/2)")

RegexMatch("100%")

In [25]:
m[1]

BoundsError: BoundsError: attempt to access 0-element Array{Union{Nothing, SubString{String}},1} at index [1]

In [26]:
names(m)

MethodError: MethodError: no method matching names(::RegexMatch)
Closest candidates are:
  names(!Matched::DataFrames.Index) at /opt/julia/packages/DataFrames/cdZCk/src/other/index.jl:34
  names(!Matched::Module; all, imported) at reflection.jl:98
  names(!Matched::DataFrames.SubIndex) at /opt/julia/packages/DataFrames/cdZCk/src/other/index.jl:421
  ...

This looked so simple, but how to get our wonderful match out of the regex object. Subsetting did not work, and trying to be fancy interrogating the object with `names` similar to what I love to do in Python with `dir`, threw an error.

Anyways, maybe we can do something else, using, say, `matchall`.

In [27]:
m2 = matchall(ptn, "100% (2/2)")

UndefVarError: UndefVarError: matchall not defined

OK, the function found on stackoverflow isn't defined, maybe because the post was from 2014. New languages are a bitch. It seems the `RegexMatch` object has methods like `match` and `captures` and `offset`.

In [28]:
m.match

"100%"

In [29]:
m.captures

0-element Array{Union{Nothing, SubString{String}},1}

In [30]:
m.offset

1

OK, whatevs, match is what I want for now. But seriously, if matchall isn't a function, what do you use when you want to capture more than one result? There can be two things in a string.

So we want to assign a new variable using an existing variable. This takes two things (in addition to the function): Accessing the existing variable, and assigning the new one.

In [31]:
first(df.AssignmentDurationInSeconds)

120

In [32]:
df.newvariable = "abcd"

┌ Warning: `setproperty!(df::DataFrame, col_ind::Symbol, v)` is deprecated, use `df[!, col_ind] .= v` instead.
│   caller = top-level scope at In[32]:1
└ @ Core In[32]:1


"abcd"

Hmmm... interesting, some new syntax here. The `.=` is peculiar. Let's try.

In [33]:
df[!, "newvar"] .= "abc"

1760-element Array{String,1}:
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 ⋮
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"
 "abc"

In [34]:
names(df)

33-element Array{String,1}:
 "HITId"
 "HITTypeId"
 "Title"
 "Description"
 "Keywords"
 "Reward"
 "CreationTime"
 "MaxAssignments"
 "RequesterAnnotation"
 "AssignmentDurationInSeconds"
 "AutoApprovalDelayInSeconds"
 "Expiration"
 "NumberOfSimilarHITs"
 ⋮
 "RejectionTime"
 "RequesterFeedback"
 "WorkTimeInSeconds"
 "LifetimeApprovalRate"
 "Last30DaysApprovalRate"
 "Last7DaysApprovalRate"
 "Input.document_url"
 "Answer.category.label"
 "Approve"
 "Reject"
 "newvariable"
 "newvar"

So far so good. Let's try putting it together.

In [35]:
[eltype(col) for col = eachcol(df)]

33-element Array{Type,1}:
 String
 String
 String
 String
 String
 String
 String
 Int64
 String
 Int64
 Int64
 String
 Missing
 ⋮
 Missing
 Missing
 Int64
 String
 String
 String
 String
 String
 Missing
 Missing
 String
 String

In [36]:
df[!, "pctapproval"] .= match.(r"^[0-9]{1,3}\%", df[!, "LifetimeApprovalRate"]).match

ErrorException: type Array has no field match

So, probably got some broadcasting issue going on here, thought that was what the dot after the function was for, but whatever. Googling how to assign variable in julia should return something, I presume.

OK lets take a step back:

In [37]:
#Pkg.add("DataFramesMeta")

In [38]:
using DataFramesMeta

In [39]:
df2 = @linq df |>
    @transform(worktimeinminutes = :LifetimeApprovalRate ./ 60)

MethodError: MethodError: no method matching transform(::Nothing; worktimeinminutes=var"#3#4"())
Closest candidates are:
  transform(!Matched::AbstractDataFrame; kwargs...) at /opt/julia/packages/DataFramesMeta/c67UK/src/DataFramesMeta.jl:377
  transform(!Matched::GroupedDataFrame; kwargs...) at /opt/julia/packages/DataFramesMeta/c67UK/src/DataFramesMeta.jl:385

It's fun to copy-paste internet code because internet code is the best code. Somehow internet code gave errors though. But don't worry, we found more internet code! This time functional, which is much promising.

In [40]:
df.somethingsomething = map( x -> match(r"^[0-9]{1,3}\%", x).match, df[!,:LifetimeApprovalRate] )

1760-element PooledArrays.PooledArray{SubString{String},UInt32,1,Array{UInt32,1}}:
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 ⋮
 "0%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "0%"
 "100%"
 "0%"
 "100%"
 "100%"

OK apparantly that works now when I didn't try to somehow pass the regex in as y, so yeab basically a lambda function because it works. But I'll own up to it, I did have a few other trys.

In [41]:
df.newstuff = match.(r"^[0-9]{1,3}\%", df[!, :LifetimeApprovalRate]).match

ErrorException: type Array has no field match

In [42]:
[thing.match for thing in match.(r"^[0-9]{1,3}\%", ["4%blah", "5%blah"])]

2-element Array{SubString{String},1}:
 "4%"
 "5%"

OK, we could probably use a list comprehension, to apply the match property on the match function, but better to just make our own short function

In [43]:
function getrx(txt)
    match(r"^[0-9]{1,3}\%", txt).match
end

df.anotherthing = getrx.(df[!, :LifetimeApprovalRate])

1760-element Array{SubString{String},1}:
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 ⋮
 "0%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "100%"
 "0%"
 "100%"
 "0%"
 "100%"
 "100%"

In [44]:
df.anotherthing[1:3]

3-element Array{SubString{String},1}:
 "100%"
 "100%"
 "100%"

Really wanted to use `first` or `head`, but apparantly that doesn't work on singe columns. Luckily, I totally forgot that a column is an array and I could just have subsetted it like normal humans.

At least you didn't have to sit and watch me. Let's just say I deleted some cells.

So OK we have created a new variable that's not just something trivial like `a+1` or something. Let's try the aggregation again.

In [57]:
df2 = @pipe df |>
    groupby(_, :anotherthing) |>
    combine(_, :WorkTimeInSeconds => mean, :LifetimeApprovalRate => length)

2×3 DataFrame
│ Row │ anotherthing │ WorkTimeInSeconds_mean │ LifetimeApprovalRate_length │
│     │ SubString…   │ Float64                │ Int64                       │
├─────┼──────────────┼────────────────────────┼─────────────────────────────┤
│ 1   │ 100%         │ 26.5624                │ 1259                        │
│ 2   │ 0%           │ 33.7305                │ 501                         │

Lastly, we want to write the data out again. And because we are civilized people we desperately don't want to stoop to the level of writing CSVs. Julia has a HDF5 library, and HDF5 is the best file format you haven't heard about.

In [47]:
using Pkg

In [48]:
Pkg.add("HDF5")

   Updating registry at `/opt/julia/registries/General`


   Updating git-repo `https://github.com/JuliaRegistries/General.git`


  Resolving package versions...
   Updating `/opt/julia/environments/v1.4/Project.toml`
 [no changes]
   Updating `/opt/julia/environments/v1.4/Manifest.toml`
 [no changes]


In [62]:
using HDF5

In [68]:
HDF5::h5write("./summary.h5", "data", df2)

#h5open("summary.h5", "w") do file
#    write(file, "data", df2)
#end

MethodError: MethodError: no method matching write(::HDF5File, ::String, ::DataFrame)
Closest candidates are:
  write(::Union{HDF5File, HDF5Group}, ::String, ::Any, !Matched::String, !Matched::Any, !Matched::Any...) at /opt/julia/packages/HDF5/ECBXn/src/HDF5.jl:1649
  write(!Matched::AbstractString, ::Any, ::Any...) at io.jl:367
  write(!Matched::IO, ::Any, ::Any...) at io.jl:584
  ...

This is priceless. MethodError and stacktrace all pretty much screaming out loud that nah, this is not working. But I'm finding a summary.h5 file, with a size greater than zero. Let's check.

In [71]:
dfread = h5read("summary.h5", "data")

HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 0:
  #000: H5O.c line 120 in H5Oopen(): unable to open object
    major: Object header
    minor: Can't open object
  #001: H5Oint.c line 596 in H5O__open_name(): unable to open object
    major: Object header
    minor: Can't open object
  #002: H5Oint.c line 551 in H5O_open_name(): object not found
    major: Object header
    minor: Object not found
  #003: H5Gloc.c line 422 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #004: H5Gtraverse.c line 851 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #005: H5Gtraverse.c line 627 in H5G__traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #006: H5Gloc.c line 378 in H5G__loc_find_cb(): object 'data' doesn't exist
    major: Symbol table
    minor: Object not found


ErrorException: Error opening object //data

In [72]:
first(dfread)

UndefVarError: UndefVarError: dfread not defined

Damn it. OK I'm a pragmatist, not messing around with julia anytime soon. We have two candidates left before submitting to the csv-hellhole: feather and parquet.

Let's do Parquet first, because Parquet is becoming fairly popular. The Parquet library is named... Parquet.

In [73]:
Pkg.add("Parquet")

  Resolving package versions...
  Installed CodecZlib ─────────────── v0.7.0
  Installed Zstd_jll ──────────────── v1.4.5+1
  Installed MemPool ───────────────── v0.2.0
  Installed LittleEndianBase128 ───── v0.3.0
  Installed CodecZstd ─────────────── v0.7.0
  Installed Snappy ────────────────── v0.3.0
  Installed Parquet ───────────────── v0.6.1
  Installed TranscodingStreams ────── v0.9.5
  Installed ThriftJuliaCompiler_jll ─ v0.11.0+0
  Installed Thrift ────────────────── v0.7.0
######################################################################### 100.0%##O#- #                                                                        
######################################################################### 100.0%##O=#  #                                                                       
   Updating `/opt/julia/environments/v1.4/Project.toml`
  [626c502c] + Parquet v0.6.1
   Updating `/opt/julia/environments/v1.4/Manifest.toml`
  [944b1d66] + CodecZlib v0.7.0
  [6b39b394] + Cod

In [74]:
using Parquet

┌ Info: Precompiling Parquet [626c502c-15b0-58ad-a749-f091afb673ae]
└ @ Base loading.jl:1260


In [75]:
write_parquet("summary.parquet", df2)

String: "Column whose `eltype` is SubString{String} is not supported at this stage. \n"

Ah OK cool, this is not really surprising when I think about it, the first column is still that weird substring datatype and naturally parquet is not participating in the weirdness. Let's try to coerce it.

In [78]:
df2.anotherthing .= string(df2.anotherthing)

2-element Array{String,1}:
 "[\"SubString{String}[\\\"100%\\\", \\\"0%\\\"]\", \"SubString{String}[\\\"100%\\\", \\\"0%\\\"]\"]"
 "[\"SubString{String}[\\\"100%\\\", \\\"0%\\\"]\", \"SubString{String}[\\\"100%\\\", \\\"0%\\\"]\"]"

Ah, now I've done it, totally ruined the column. OK on the one hand I really want to fix it, on the other hand I really don't care because what I wanted to do was to write this file. Let's do the last first, and maybe do a little substring type-exploration later.

In [93]:
write_parquet("summary.parquet", df2)

OK, so far so good. Manual inspections says the file is there as well. Let's try to read it again.

In [94]:
summary_parquet = ParFile("summary.parquet")

Parquet file: summary.parquet
    version: 1
    nrows: 2
    created by: Parquet.jl 0.6.0
    cached: 0 column chunks


In [95]:
first(summary_parquet)

MethodError: MethodError: no method matching iterate(::ParFile)
Closest candidates are:
  iterate(!Matched::Core.SimpleVector) at essentials.jl:603
  iterate(!Matched::Core.SimpleVector, !Matched::Any) at essentials.jl:603
  iterate(!Matched::ExponentialBackOff) at error.jl:253
  ...

In [96]:
summary_parquet |> DataFrame

ArgumentError: ArgumentError: no default `Tables.columns` implementation for type: ParFile

OK so this becomes something other than a dataframe. I'm sure it's possible to beat it into submission, but I seriously don't have the patience for that. Let's try feather instead. There is a Feather library it seems, it looks a lot younger, but the little documentation I found does suggest it hasn't tried to create its own dataframe format. In other words, possibly less nihilistic.

In [97]:
Pkg.add("Feather")

  Resolving package versions...
  Installed Arrow ─────── v0.2.4
  Installed FlatBuffers ─ v0.5.4
  Installed Feather ───── v0.5.6
   Updating `/opt/julia/environments/v1.4/Project.toml`
  [becb17da] + Feather v0.5.6
   Updating `/opt/julia/environments/v1.4/Manifest.toml`
  [69666777] + Arrow v0.2.4
  [becb17da] + Feather v0.5.6
  [53afe959] + FlatBuffers v0.5.4


In [98]:
using Feather

┌ Info: Precompiling Feather [becb17da-46f6-5d3c-ad1b-1c5fe96bc73c]
└ @ Base loading.jl:1260


In [99]:
Feather.write("summary.feather", df2)

"summary.feather"

In [100]:
summary_feather = Feather.read("summary.feather")

2×3 DataFrame. Omitted printing of 3 columns
│ Row │ │     │ ├─────┼
│ 1   │ │ 2   │

In [101]:
first(summary_feather)

DataFrameRow. Omitted printing of 3 columns
│ Row │ │     │ ├─────┼
│ 1   │

In [102]:
names(summary_feather)

3-element Array{String,1}:
 "anotherthing"
 "WorkTimeInSeconds_mean"
 "LifetimeApprovalRate_length"

In [104]:
summary_feather.WorkTimeInSeconds_mean

2-element Arrow.Primitive{Float64}:
 26.562351072279586
 33.73053892215569

OK, it looks like we found a winner.

**As promised, a small foray into the weird substring datatype**

In [81]:
sbst = df.anotherthing[1:3]

3-element Array{SubString{String},1}:
 "100%"
 "100%"
 "100%"

In [84]:
typeof(sbst[1])

SubString{String}

In [85]:
sst = sbst[1]

"100%"

In [105]:
sst.string

"100% (5/5)"

In [106]:
sst.offset

0

In [107]:
sst.ncodeunits

4

This is really weird, there only 3 methods that I can find and no "just get the damn text" method, but it is happy to return the entire string (not the subsstring) and some offsets even 0-indexed so when doing substring stuff I have to add 1. Still when printing this shit, it gives me the actual, real value. There has to be a better way, but...

In [108]:
sst.string[sst.offset+1:sst.ncodeunits]

"100%"

OK, We're going to stop there for now, let's take a look at visualizations next. At the end, maybe we can even combine dataframes and visualizations, so that we are visualizing data we read from a CSV file. Or some other type of file.